<a href="https://colab.research.google.com/github/alina-bogdanovskaya/Linear_Algebra_Feb2022/blob/lesson5/SVD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

##__1.__ Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

In [14]:
a = np.array([[1,2,0],[0,0,5],[3,-4,2],[1,6,5],[0,1,0]],float)

In [15]:
U,s,W = np.linalg.svd(a)
U

array([[ 0.17056501,  0.15680918, -0.53077508, -0.79905375, -0.16158397],
       [ 0.39287016, -0.52933945,  0.6134793 , -0.43375771,  0.03082495],
       [-0.14366152, -0.82449256, -0.52379105,  0.14049848,  0.07400343],
       [ 0.88843702,  0.06074346, -0.24655277,  0.37755832, -0.06042632],
       [ 0.08125046,  0.10831843, -0.08231425, -0.10524851,  0.98173958]])

In [16]:
U.T@U

array([[ 1.00000000e+00,  3.19674802e-16, -3.99355622e-16,
         1.35663262e-17, -8.83091830e-18],
       [ 3.19674802e-16,  1.00000000e+00,  3.63628026e-16,
         7.29367646e-17,  3.15830273e-17],
       [-3.99355622e-16,  3.63628026e-16,  1.00000000e+00,
        -1.56755518e-16, -7.11991378e-17],
       [ 1.35663262e-17,  7.29367646e-17, -1.56755518e-16,
         1.00000000e+00, -2.63859021e-17],
       [-8.83091830e-18,  3.15830273e-17, -7.11991378e-17,
        -2.63859021e-17,  1.00000000e+00]])

In [17]:
V = W.T
V

array([[ 0.07116451, -0.36737824, -0.92734505],
       [ 0.71702467,  0.66514082, -0.20847855],
       [ 0.69340553, -0.65009301,  0.31075368]])

In [18]:
V.T@V

array([[ 1.00000000e+00,  2.63580532e-16, -8.46021341e-17],
       [ 2.63580532e-16,  1.00000000e+00, -1.38296348e-16],
       [-8.46021341e-17, -1.38296348e-16,  1.00000000e+00]])

In [19]:
D = np.zeros_like(a, dtype=float)
D[np.diag_indices(min(a.shape))] = s
D

array([[8.82486885, 0.        , 0.        ],
       [0.        , 6.14060608, 0.        ],
       [0.        , 0.        , 2.53271528],
       [0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        ]])

In [24]:
U@D@V.T

array([[ 1.00000000e+00,  2.00000000e+00,  1.31018987e-15],
       [ 1.93009612e-16, -2.66609409e-15,  5.00000000e+00],
       [ 3.00000000e+00, -4.00000000e+00,  2.00000000e+00],
       [ 1.00000000e+00,  6.00000000e+00,  5.00000000e+00],
       [-5.49534701e-17,  1.00000000e+00, -3.02821608e-17]])

##__2.__ Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

а) $\left \| A \right \|_{E}=\mu_{1} = 8.82486885 $

In [34]:
np.linalg.norm(a,2)

8.824868854820444

б) $\left \| A \right \|_{F}= \sqrt{1+4+25+9+16+4+1+36+25+1} = \sqrt{122} \approx 11.0454$

In [33]:
np.linalg.norm(a,'fro')

||A||_F = 11.045361017187261
